Provide implementation of `selectCoins` API function.

jonathanknowles opened the pull request:
# Issue Number

#1136 

# Overview

This PR:

- [x] Provides an implementation of the `selectCoins` API function.
- [x] Provides an integration test to verify that a singleton payment is included in the output.
- [x] Provides an integration test to verify that multiple payments are included in the output.
- [x] Adds an `UnsignedTx` type to capture the concept of a set of inputs and outputs that are ready to be signed into a real `Tx`.

jonathanknowles committed:
Add function `Wallet.assignChangeAddressesAndCheckpoint`.

This function acts as a wrapper around `assignChangeAddresses`, with the
additional responsibility of writing a database checkpoint to reflect
the updated wallet state.

jonathanknowles committed:
Add function `Api.Server.mkApiCoinSelection`.

jonathanknowles committed:
Provide implementation of `Api.Server.selectCoins`.

jonathanknowles committed:
Add boilerplate testing functions to access fields from `ApiCoinSelection`.

jonathanknowles committed:
Extend `WALLETS_COIN_SELECTION_01` to check that a singleton payment is included.

jonathanknowles committed:
Add `WALLETS_COIN_SELECTION_02` to check that multiple payments are included.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -985,6 +986,43 @@ signPayment ctx wid argGenChange pwd (CoinSelection ins outs chgs) = db & \DBLay
+--
+-- See 'assignChangeAddresses'.
+--
+assignChangeAddressesAndCheckpoint
```
**[change]**

Rename this to (for example) `selectCoinsForExternal` and move the logic from the API handler to here.

**Why?** 

Having to define errors type for the Api handlers is a good sign that some logic is in the wrong place. We should keep the API handlers simple and move as much as the wallet logic down to this module. The API layer should only compose functions from the wallet engine. Plus, the name `assignChangeAddressesAndCheckpoint` already suggest that something seems wrong (everytime one is tempted to put `And` in the name of a function is a good sign that this function is probably not operating at the right level of abstraction).

jonathanknowles replied:
> Rename this to (for example) selectCoinsForExternal and move the logic from the API handler to here.

Fair enough! I've made the following change to implement your suggestion: https://github.com/input-output-hk/cardano-wallet/pull/1165/commits/62b45c411171f5e91ea044f50e48144596faf78b.

Of course, let me know if this is not what you had in mind.

As a result of the above change, the tuple `(NonEmpty (TxIn, TxOut), NonEmpty TxOut)` appears in the signatures of both the following functions:

* `selectCoinsForExternal`
* `mkApiCoinSelection`

To simplify this, I've made a further commit which replaces this tuple with a dedicated type, to capture the concept of an _**unsigned transaction**_: https://github.com/input-output-hk/cardano-wallet/pull/1165/commits/09ce374339f33cb9f711ff7804738313f1bb3c82.

Let me know if this is suitable. If not, I'd be happy to revise it further. Thanks in advance!

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -622,14 +625,43 @@ coinSelections =
+        <*> (ensureNonEmptyOutputs =<<
+                liftHandler (withWorkerCtx ctx wid throwE $ \wrk ->
+                    W.assignChangeAddressesAndCheckpoint
+                        @_ @s @k wrk wid () mOutputs mChange))
```
**[suggestion]**

Swap the use of applicative for a monadic approach. 

```hs
let payments = coerceCoin <$> (body ^. #payments)
CoinSelection mInputs mRecipients mChange <- liftHandler 
    $ withWorkerCtx ctx wid liftE 
    $ \wrk -> W.selectCoinsForPayment @_ @s @t wrk wid payments

mOutputs  <- liftHandler 
    $ withWorkerCtx ctx wid throwE 
    $ \wrk -> assignChangeAddressesAndCheckpoint @_ @s @k wrk wid () mRecipients mChange

neInputs  <- ensureNonEmpty ErrSelectCoinsUnableToAssignInputs mInputs
neOutputs <- ensureNonEmpty ErrSelectCoinsUnableToAssignOutputs mOutputs

pure $ mkApiCoinSelection neInputs neOutputs

```

**Why?**

(In this case) It leads to more readable code without changing much of the semantic.

jonathanknowles replied:
Hi @KtorZ. Thanks for this suggestion: I agree this could be more readable.

I've made the following change, in response to your suggestion: https://github.com/input-output-hk/cardano-wallet/pull/1165/commits/c9a6b065b733612b876860ed2b2aea928fd7b549.

jonathanknowles committed:
Simplify applicative expression in `selectCoins`.

jonathanknowles committed:
Push logic from `Server.selectCoins` down into the wallet layer.

jonathanknowles committed:
Introduce type `UnsignedTx`.

Use this type to:

1. simplify the type of `Wallet.selectCoinsExternal`;
2. simplify the type of `Server.mkApiCoinSelection`.
3. simplify function `Server.selectCoins`;

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -985,6 +988,55 @@ signPayment ctx wid argGenChange pwd (CoinSelection ins outs chgs) = db & \DBLay
+                pure mOutputs
+    UnsignedTx
+        <$> ensureNonEmpty mInputs  ErrSelectCoinsExternalUnableToAssignInputs
+        <*> ensureNonEmpty mOutputs ErrSelectCoinsExternalUnableToAssignOutputs
```
**[remark]** 

I am not fond of having `xxxUnableToAssignxxx` at the value-level here because, these are clearly invariant violation. If you give a non-empty list of output to the coin selection (as we do) then the result MUST have non empty inputs and outputs. Anything else is an invariant violation and ought to be treated as an exceptional case (i.e. with an exception). 

In a sense, this is a consequence of our "insufficient" typing regarding `CoinSelection`. A possible "easy fix" would be to simply move that contract inside the coin selection functions (`random` and `largestFirst`) adding invariant about the shape of the result.

jonathanknowles replied:
> I am not fond of having `xxxUnableToAssignxxx` at the value-level here because, these are clearly invariant violation. If you give a non-empty list of output to the coin selection (as we do) then the result MUST have non empty inputs and outputs. Anything else is an invariant violation and ought to be treated as an exceptional case (i.e. with an exception).
> 
> In a sense, this is a consequence of our "insufficient" typing regarding `CoinSelection`. A possible "easy fix" would be to simply move that contract inside the coin selection functions (`random` and `largestFirst`) adding invariant about the shape of the result.

I completely agree, and think it would be better to have the "cannot be empty" invariants encoded in the `CoinSelection` type itself.

I'm guessing this would require quite a bit of refactoring. Perhaps we should put this on our list to tidy up during a recovery week?
